clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java

```diff
@@ -486,6 +487,43 @@ public void testRetainMemberIdAfterSyncGroupDisconnect() {
         ensureActiveGroup(rejoinedGeneration, memberId);
     }
 
+    @Test
+    public void testRejoinReason() {
+        setupCoordinator();
+
+        String memberId = "memberId";
+        int generation = 5;
+
+        // test initial reason
+        mockClient.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
+        expectJoinGroup("", "initialized abstract coordinator", generation, memberId);
+
+        // successful sync group response should reset reason
+        expectSyncGroup(generation, memberId);
+        ensureActiveGroup(generation, memberId);
+        assertEquals("", coordinator.rejoinReason());
+
+        // Force a rebalance
+        expectJoinGroup(memberId, "Manual test trigger", generation, memberId);
+        expectSyncGroup(generation, memberId);
+        coordinator.requestRejoin("Manual test trigger");
+        ensureActiveGroup(generation, memberId);
+        assertEquals("", coordinator.rejoinReason());
+
+        // max group size reached
+        mockClient.prepareResponse(joinGroupFollowerResponse(defaultGeneration, memberId, JoinGroupRequest.UNKNOWN_MEMBER_ID, Errors.GROUP_MAX_SIZE_REACHED));
+        coordinator.requestRejoin("Manual test trigger 2");
+        try {
+            coordinator.joinGroupIfNeeded(mockTime.timer(100L));
+        } catch (GroupMaxSizeReachedException e) {
```

Comment:
any reason why we don't use `assertThrows` here?

Reply:
we do not want to stop when an exception is thrown - we need to verify that the reason is updated to the appropriate exception message.

Reply:
Actually, you can achieve what you want by `assertThrows` as below:

```java
        Throwable e = assertThrows(GroupMaxSizeReachedException.class,
            () -> coordinator.joinGroupIfNeeded(mockTime.timer(100L)));

        // next join group request should contain exception message
        expectJoinGroup(memberId, e.getMessage(), generation, memberId);
        expectSyncGroup(generation, memberId);
        ensureActiveGroup(generation, memberId);
        assertEquals("", coordinator.rejoinReason());
```

Basically, the `assertThrows` is doing the similar thing as what you did here (try/catch). It's recommended to use the `assertThrows` to understand the exception is also a verification position in this test. Otherwise, let's say, if someday, someone breaks the logic and makes the `coordinator.joinGroupIfNeeded(mockTime.timer(100L));` works without exception thrown, your current test can't catch this error, right?

Reply:
ah gotcha. makes sense, thanks!